### PR TITLE
bpo-42504: fix for MACOSX_DEPLOYMENT_TARGET=11

### PR DESCRIPTION
**Pull request by fxcoudert**
This pull request provides a fix for https://bugs.python.org/issue42504

https://bugs.python.org/issue42504

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@fxcoudert

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by fxcoudert** [comment]
I have now signed the CLA

**Review by guillochon — APPROVED** [review]

**Comment by lewisrobbins** [comment]
Looks good to me 👍

**Review by ned-deily — CHANGES_REQUESTED** [review]
Thanks for the PR. There is more work to be done, though. With:

`MACOSX_DEPLOYMENT_TARGET=11 ./configure
`

making this change in `setup.py` gets past the first error only to fail later in `distutils/spawn.py`: 

```
  File "source/Lib/distutils/spawn.py", line 66, in spawn
    if _cfg_target_split > [int(x) for x in cur_target.split('.')]:
TypeError: '>' not supported between instances of 'NoneType' and 'list'
```
 Are there other users of `MACOSX_DEPLOYMENT_TARGET` in the source that need to be changed as well?

**Comment by ronaldoussoren** [comment]
A nicer fix is to ensure that sysconfig.get_config_var('MACOSX_DEPLOYMENT_TARGET') evaluates to a string. I haven't checked how easy it would be to do this.

**Comment by fxcoudert** [comment]
@ronaldoussoren the code at https://github.com/python/cpython/blob/3.9/Lib/sysconfig.py#L461 shows it's very deliberate that int-type values are returned as int

**Comment by fxcoudert** [comment]
Two places I am not 100% sure what to do:
- https://github.com/python/cpython/blob/master/configure.ac#L1970 this assumes that `cur_target="${cur_target_major}.${cur_target_minor}"`, so it may set `MACOSX_DEPLOYMENT_TARGET` to `11.0`. I think that's acceptable.
- https://github.com/python/cpython/blob/master/Mac/BuildScript/build-installer.py has lots of use cases, but I think they're all string-based, and therefore fine

**Comment by fxcoudert** [comment]
I have made the requested changes; please review again.

Travis is telling me:

```
Fixing Python file whitespace ... 1 file:
  Lib/test/test_posix.py
```

but I haven't changed whitespace in this file, and it is not telling me what the issue actually is.

File: Lib/distutils/spawn.py [hunk 1]
```diff
@@ -57,7 +57,7 @@ def spawn(cmd, search_path=1, verbose=0, dry_run=0):
             _cfg_target = sysconfig.get_config_var(
                                   'MACOSX_DEPLOYMENT_TARGET') or ''
             if _cfg_target:
-                _cfg_target_split = [int(x) for x in _cfg_target.split('.')]
+                _cfg_target_split = [int(x) for x in str(_cfg_target).split('.')]
```
Comment (ned-deily) [inline Lib/distutils/spawn.py]: A similar change is needed below at line 66.

**Review by ned-deily — CHANGES_REQUESTED** [review]
Thanks for the changes. Running test_distutils:
`./python(.exe) -m test -w test_distutils`

gives a number of failing test cases, basically due to one of two problems.

1. The missing `str()` at spawn.py:66

2. Failures in `test_build_ext.py:501`:
```
File "source/Lib/distutils/tests/test_build_ext.py", line 501, in _try_compile_deployment_target
    target = '%02d%02d00' % target
TypeError: not enough arguments for format string
```

I don't think we need to make any changes to `configure.ac` or `build_installer.py`.

**Comment by fxcoudert** [comment]
I have made the requested changes; please review again.
There is still a whitespace failure about `Lib/test/test_posix.py`, but I haven't changed whitespace. I've ran `make patchcheck` but it fails, maybe I'm not doing it properly.

**Comment by nilleb** [comment]
I think there is one more occurrence to fix at `subprocess.py` line 1739: (👇 this is the fixed version)
```
                                val = os.fsencode(str(v))
```
because `v` is an environment variable value, and integers are not acceptable input values for os.fsencode

**Comment by fxcoudert** [comment]
@nilleb I don't think this is a problem, environment variables are strings in all cases. This is unrelated to the problem of `sysconfig.get_config_var`, which does return str or int depending on the content.

**Comment by fxcoudert** [comment]
@nilleb wait I see the problem you're referring to, we're passing that int as an environment variable. That's where it should be fixed, actually. If it's defined, `_cfg_target` should always be a string to begin with. Then things go smoothly.

**Comment by fxcoudert** [comment]
This was tested with a full build of Python on macOS Big Sur (ARM), as well as building several python-based packages. Without the latest version of the patch, some failed in `subprocess.py`. Now all pass.

**Comment by nilleb** [comment]
hey @fxcoudert  the problem comes from line https://github.com/python/cpython/pull/23556/files#diff-a06590fdcc4068035ba08a86c9c816cb70a78d80cb9ff0babfbb8dae85dd41d0L75
where we pass `env=env` to the subprocess call (this has nothing to do with `_cfg_target`)
Since the variable `MACOSX_DEPLOYMENT_TARGET` has a `11` (`int`)  value
3 stack levels beneath, it will raise an exception.

you can reproduce it by

```sh
git clone https://github.com/nilleb/magpie
cd magpie
bin/setup
```

**Comment by fxcoudert** [comment]
@nilleb this is fixed in the latest version of this patch. The problem is we pass `cur_target` as `MACOSX_DEPLOYMENT_TARGET`, and `cur_target` defaults to `_cfg_target` in some cases, and `_cfg_target` could be an int. Now it's always a str, and there is no problem.

**Comment by guillochon** [comment]
@fxcoudert can you add the blurb using the blurb-it tool so that check will pass? https://blurb-it.herokuapp.com/

**Comment by guillochon** [comment]
@ned-deily mind taking a gander at this one again? It looks ready to go

**Review by ned-deily — APPROVED** [review]
LGTM now, thanks!  No doubt we will find other implications of moving from 10.x to 1x release version but this should be a good start.

**Comment by fxcoudert** [comment]
Related PR, with `wheel`: https://github.com/pypa/wheel/issues/385

**Pull request merged**